pkg/finality-grandpa/bitfield.go

```diff
@@ -0,0 +1,158 @@
+// at the specified position, it is resized accordingly.
+func (b *Bitfield) SetBit(position uint) {
+	wordOff := position / 64
+	bitOff := position % 64
+
+	if wordOff >= uint(len(b.bits)) {
+		newLen := wordOff + 1
+		b.bits = append(b.bits, make([]uint64, newLen-uint(len(b.bits)))...)
+	}
+	b.bits[wordOff] |= 1 << (63 - bitOff)
+}
+
+// Get an iterator over all bits that are set (i.e. 1) in the bitfield,
+// starting at bit position `start` and moving in steps of size `2^step`
+// per word.
+func (b Bitfield) iter1s(start, step uint) (bit1s []Bit1) {
+	return iter1s(b.bits, start, step)
+}
+
+// Get an iterator over all bits that are set (i.e. 1) at even bit positions.
+func (b Bitfield) Iter1sEven() []Bit1 {
+	return b.iter1s(0, 1)
+}
+
+// Get an iterator over all bits that are set (i.e. 1) at odd bit positions.
+func (b Bitfield) Iter1sOdd() []Bit1 {
+	return b.iter1s(1, 1)
+}
+
+// Get an iterator over all bits that are set (i.e. 1) when merging
+// this bitfield with another bitfield, without modifying either
+// bitfield, starting at bit position `start` and moving in steps
+// of size `2^step` per word.
+func (b Bitfield) iter1sMerged(other Bitfield, start, step uint) []Bit1 {
+	switch {
+	case len(b.bits) == len(other.bits):
+		zipped := make([]uint64, len(b.bits))
+		for i, a := range b.bits {
+			b := other.bits[i]
+			zipped[i] = a | b
+		}
+		return iter1s(zipped, start, step)
+	case len(b.bits) < len(other.bits):
+		zipped := make([]uint64, len(other.bits))
+		for i, bit := range other.bits {
+			var a uint64
+			if i < len(b.bits) {
+				a = b.bits[i]
+			}
+			zipped[i] = a | bit
+		}
+		return iter1s(zipped, start, step)
+	case len(b.bits) > len(other.bits):
+		zipped := make([]uint64, len(b.bits))
+		for i, a := range b.bits {
+			var b uint64
+			if i < len(other.bits) {
+				b = other.bits[i]
+			}
+			zipped[i] = a | b
+		}
+		return iter1s(zipped, start, step)
+	default:
+		panic("unreachable")
+	}
+}
+
+// Get an iterator over all bits that are set (i.e. 1) at even bit positions
+// when merging this bitfield with another bitfield, without modifying
+// either bitfield.
+func (b Bitfield) Iter1sMergedEven(other Bitfield) []Bit1 {
+	return b.iter1sMerged(other, 0, 1)
+}
+
+// Get an iterator over all bits that are set (i.e. 1) at odd bit positions
+// when merging this bitfield with another bitfield, without modifying
+// either bitfield.
+func (b Bitfield) Iter1sMergedOdd(other Bitfield) []Bit1 {
+	return b.iter1sMerged(other, 1, 1)
+}
+
+// Turn an iterator over u64 words into an iterator over bits that
+// are set (i.e. `1`) in these words, starting at bit position `start`
+// and moving in steps of size `2^step` per word.
+func iter1s(iter []uint64, start, step uint) (bit1s []Bit1) {
+	if !(start < 64 && step < 7) {
+		panic("wtf?")
+	}
+	steps := (64 >> step) - (start >> step)
+	for i, word := range iter {
+		n := steps
+		if word == 0 {
+			n = 0
+		}
+		for j := uint(0); j < n; j++ {
```

Comment:
actually you don't need the `n` variable, if the `word == 0` you can jump to the next one otherwise exec the inner for loop using the `steps` as a control variable
```suggestion
		if word == 0 {
			continue
		}
		for j := uint(0); j < steps; j++ {
```